libr/util/logging.c

```diff
@@ -0,0 +1,123 @@
+/* radare - LGPL - Copyright 2007-2013 - pancake */
+
+#define LOGGING_CONFIGSTR_SIZE 512
+#define LOGGING_OUTPUTBUF_SIZE 512
+
+#include <r_core.h>
+#include <stdarg.h>
+
+// TODO: Use thread-local storage to make these variables thread-safe
+static RLoggingFuncdef cb_logfunc_print = NULL; // Function to call when outputting log string
+static int cfg_loglvl = R_LOGLVL_ERROR; // Logging level output
+static int cfg_logtraplvl = R_LOGLVL_FATALTRAP; // Logging trap level
+static bool cfg_logsrcinfo = false; // Print out debug source info with the output
+static bool cfg_logcolors = false; // Output colored log text based on level
+static char cfg_logfile[LOGGING_CONFIGSTR_SIZE] = ""; // Output text to filename
+static const char *level_tags[] = { // Log level to tag string lookup array
+	[R_LOGLVL_SILLY]     = "SILLY: ",
+	[R_LOGLVL_VERBOSE]   = "VERBOSE: ",
+	[R_LOGLVL_DEBUG]     = "DEBUG: ",
+	[R_LOGLVL_INFO]      = "INFO: ",
+	[R_LOGLVL_WARN]      = "WARNING: ",
+	[R_LOGLVL_ERROR]     = "ERROR: ",
+	[R_LOGLVL_FATAL]     = "FATAL: ",
+	[R_LOGLVL_FATALTRAP] = "FATAL: "
+};
+
+/* TODO: Might want to create a r_logging_use_envvars function so that programs
+   that use r_util without core/cons have some way of changing the above vars
+   to the environmental variables. */
+
+// cconfig.c configuration callback functions below
+void r_logging_set_level(RLoggingLevel level) {
+	cfg_loglvl = level;
+}
+
+void r_logging_set_traplevel(RLoggingLevel level) {
+	cfg_logtraplvl = level;
+}
+
+void r_logging_set_file(const char *filename) {
+	int value_len = strnlen (filename, LOGGING_CONFIGSTR_SIZE) + 1;
+	strncpy (&cfg_logfile, filename, value_len);
+}
+
+void r_logging_set_srcinfo(bool show_info) {
+	cfg_logsrcinfo = show_info;
+}
+
+void r_logging_set_colors(bool show_info) {
+	cfg_logcolors = show_info;
+}
+
+/**
+ * \brief Set the logging callback function for printing
+ * \param cbfunc RLoggingFuncdef style function to be called
+*/
+void r_logging_set_callback(RLoggingFuncdef cbfunc) {
```

Comment:
can you add `R_API` at the beginning of the definition? I'm not sure *why* we use it everywhere, but we are used to do that so just at least for consistency...